File: src/features/plugin/escopo.py
import logging
from pathlib import Path

from src.features.plugin.modelos import ScopeName, ScopeTarget

logger = logging.getLogger(__name__)


def find_project_root(start: Path) -> Path | None:
    for candidate in (start.resolve(), *start.resolve().parents):
        git_path = candidate / ".git"
        if git_path.is_dir() or git_path.is_file():
            return candidate
    return None


def resolve_scope(scope: ScopeName, start: Path, home: Path | None = None) -> ScopeTarget:
    if scope == "user":
        user_root = home if home is not None else Path.home()
        return ScopeTarget(scope=scope, root=user_root, used_git_root=False)

    project_root = find_project_root(start)
    if project_root is not None:
        return ScopeTarget(scope=scope, root=project_root, used_git_root=True)

    logger.warning("Git não encontrado; usando o diretório atual.")
    return ScopeTarget(scope=scope, root=start.resolve(), used_git_root=False)

File: src/features/plugin/escopo.ts
import { existsSync, statSync } from "node:fs";
import { homedir } from "node:os";
import { join, resolve, dirname } from "node:path";

import { logger } from "../../shared/runtime.js";
import type { ScopeName, ScopeTarget } from "./modelo.js";

export function findProjectRoot(start: string): string | null {
  let candidate = resolve(start);
  for (;;) {
    const gitPath = join(candidate, ".git");
    if (existsSync(gitPath) && (statSync(gitPath).isDirectory() || statSync(gitPath).isFile())) {
      return candidate;
    }
    const parent = dirname(candidate);
    if (parent === candidate) return null;
    candidate = parent;
  }
}

export function resolveScope(scope: ScopeName, start: string, home?: string): ScopeTarget {
  if (scope === "user") {
    const userRoot = home !== undefined ? home : homedir();
    return { scope: scope, root: userRoot, usedGitRoot: false };
  }

  const projectRoot = findProjectRoot(start);
  if (projectRoot !== null) {
    return { scope: scope, root: projectRoot, usedGitRoot: true };
  }

  logger.warn("Git não encontrado; usando o diretório atual.");
  return { scope: scope, root: resolve(start), usedGitRoot: false };
}
